src/Microsoft.AspNet.OData.Shared/EdmChangedObjectCollectionOfT.cs

```diff
@@ -0,0 +1,204 @@
+        public EdmChangedObjectCollection(IEdmEntityType entityType, IList<IEdmChangedObject<TStructuralType>> changedObjectList)
+            : base(entityType, changedObjectList as IList<IEdmChangedObject>)
+        {            
+            _items = changedObjectList;
+        }
+
+        /// <inheritdoc/>
+        public void Add(IEdmChangedObject<TStructuralType> item)
+        {
+            InsertItem(_items.Count, item);
+        }
+       
+        /// <inheritdoc/>
+        public bool Contains(IEdmChangedObject<TStructuralType> item)
+        {
+            return _items.Contains(item);
+        }
+
+        /// <inheritdoc/>
+        public void CopyTo(IEdmChangedObject<TStructuralType>[] array, int arrayIndex)
+        {
+            _items.CopyTo(array, arrayIndex);
+        }
+
+        /// <inheritdoc/>
+        public bool IsReadOnly
+        {
+            get { return false; }
+        }
+
+        /// <inheritdoc/>
+        public bool Remove(IEdmChangedObject<TStructuralType> item)
+        {
+            int count = _items.Count;
+            RemoveItem(_items.IndexOf(item));
+
+            return count == _items.Count + 1;
+        }
+
+        /// <inheritdoc/>
+        IEnumerator<IEdmChangedObject<TStructuralType>> IEnumerable<IEdmChangedObject<TStructuralType>>.GetEnumerator()
+        {
+            return _items.GetEnumerator();
+        }
+
+        /// <inheritdoc/>
+        protected override void InsertItem(int index, IEdmChangedObject item)
+        {
+            IEdmChangedObject<TStructuralType> _item = item as IEdmChangedObject<TStructuralType>;
+            if (_item == null)
+            {
+                throw Error.Argument("item", SRResources.ChangedObjectTypeMismatch, typeof(TStructuralType), item.GetType());
+            }
+
+            _items.Add(_item);
+            base.InsertItem(index, item);
+        }
+
+        /// <inheritdoc/>
+        protected override void ClearItems()
+        {
+            _items.Clear();
+        }
+
+        /// <inheritdoc/>
+        protected override void RemoveItem(int index)
+        {
+            Contract.Assert(_items.Count > index);
+            
+            _items.RemoveAt(index);
+            base.RemoveItem(index);
+        }
+
+        /// <inheritdoc/>
+        protected override void SetItem(int index, IEdmChangedObject item)
+        {
+            IEdmChangedObject<TStructuralType> _item = item as IEdmChangedObject<TStructuralType>;
+            Contract.Assert(_item != null);
+
+            _items[index] = _item;
+        }
+
+        /// <summary>
+        /// Copy changed values is an implementation of Patch
+        /// </summary>
+        /// <param name="original">Original collection of the Type which needs to be updated</param>              
+        public void CopyChangedValues(ICollection<TStructuralType> original)
```

Comment:
>void [](start = 15, length = 4)

Perhaps we should return a ChangedObjectCollection<TStructuralType> containing the applied changes -- this would be the same as the input for the ones that didn't change, and we'd change inserts to deletes, deletes to inserts, and updates to use original values for any action that failed  (as well as marking it with a DataModificationException). In the future, maybe we could even support the user supplying the custom implementations for insert/update/delete..